drivers/flash/flash_ll_nrfxlib.c

```diff
@@ -0,0 +1,313 @@
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+	.page_layout = btctlr_flash_page_layout_get,
+#endif  /* CONFIG_FLASH_PAGE_LAYOUT */
+	.write_block_size = FLASH_DRIVER_WRITE_BLOCK_SIZE
+};
+
+
+/* Utility functions */
+static inline bool is_addr_valid(off_t addr, size_t len)
+{
+	if ((addr + len) > NRF_FICR->CODEPAGESIZE * NRF_FICR->CODESIZE ||
+	    addr < 0) {
+		return false;
+	}
+
+	return true;
+}
+
+static inline bool is_aligned_32(off_t addr)
+{
+	return ((addr & 0x3) == 0);
+}
+
+static inline off_t align_32(off_t addr)
+{
+	return (addr & ~0x3);
+}
+
+static inline size_t bytes_to_words(size_t bytes)
+{
+	return bytes / sizeof(u32_t);
+}
+
+static inline bool is_page_aligned(off_t addr)
+{
+	return (addr & (NRF_FICR->CODEPAGESIZE - 1)) == 0;
+}
+
+static void flash_operation_complete_callback(u32_t status)
+{
+	__ASSERT_NO_MSG(flash_state.op == FLASH_OP_WRITE ||
+			flash_state.op == FLASH_OP_ERASE);
+	LOG_DBG("Flash op %u complete", flash_state.op);
+	if (flash_state.op == FLASH_OP_WRITE) {
+		LOG_HEXDUMP_DBG(flash_state.data, flash_state.prev_len, "wrote");
+	}
+
+	int err;
+	flash_state.addr += flash_state.prev_len;
+	flash_state.data = (const void *) ((intptr_t) flash_state.data + flash_state.prev_len);
+	flash_state.len -= flash_state.prev_len;
+
+	if (flash_state.len > 0) {
+		err = flash_op_execute();
+		/* All inputs should have been validated on the first call. */
+		__ASSERT(err == 0, "Continued flash operation failed\n");
+	} else {
+		flash_state.op = FLASH_OP_NONE;
+		k_sem_give(&flash_state.sem);
+	}
+}
+
+static size_t offset_32(const void *addr)
+{
+	return (const u32_t) addr & 0x3;
+}
+
+static size_t unaligned_word_copy(u32_t *word_dst, void *dst, const void *src, size_t len)
+{
+	/* Example: Writing one byte from src to dst via word_dst.
+	 *
+	 * |    |<----- src (len = 1)
+	 * |  |<------- dst
+	 * | 0 1 2 3 |
+	 * | address |
+	 *
+	 * max_offset             : max(1, 2) => 2
+	 * remaining_bytes_in_word: 4 - 2     => 2
+	 * bytes_to_copy          : MIN(1, 2) => 1
+	 */
+	size_t max_offset = MAX(offset_32(dst), offset_32(src));
+	size_t remaining_bytes_in_word = sizeof(u32_t) - max_offset;
+	size_t bytes_to_copy = MIN(len, remaining_bytes_in_word);
+
+	/* nRF52832 Product specification:
+	 *  Only full 32-bit words can be written to Flash using the
+	 *  NVMC interface. To write less than 32 bits to Flash, write
+	 *  the data as a word, and set all the bits that should remain
+	 *  unchanged in the word to '1'. */
+	*word_dst = ~0;
+	memcpy(&((u8_t *)word_dst)[offset_32(dst)], src, bytes_to_copy);
+	return bytes_to_copy;
+}
+
+static int flash_op_write(void)
+{
+	if (is_aligned_32(flash_state.addr) &&
+	    is_aligned_32((off_t) flash_state.data) &&
+	    flash_state.len >= sizeof(u32_t)) {
+		flash_state.prev_len = MIN(align_32(flash_state.len), NRF_FICR->CODEPAGESIZE);
+		return ble_controller_flash_write((u32_t) flash_state.addr,
+						  flash_state.data,
+						  bytes_to_words(flash_state.prev_len),
```

Comment:
Data lost risk here. Should not the bytes_to_words function ceil round the returned value? Or is the ```prev_len``` value already aligned?

Reply:
The `prev_len` is aligned on line 158 above.